contracts/strategies/BaseStrategy.sol

```diff
@@ -0,0 +1,334 @@
+        _stakeDepositTokens(_amount, depositFee);
+        emit Deposit(_account, _amount);
+    }
+
+    /**
+     * @notice Deposit fee bips from underlying farm
+     */
+    function _getDepositFeeBips() internal view virtual returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @notice Calculate deposit fee of underlying farm
+     * @dev Override if deposit fee is calculated dynamically
+     */
+    function _calculateDepositFee(uint256 _amount) internal view virtual returns (uint256) {
+        uint256 depositFeeBips = _getDepositFeeBips();
+        return (_amount * depositFeeBips) / _bip();
+    }
+
+    function withdraw(uint256 _amount) external override {
+        uint256 depositTokenAmount = getDepositTokensForShares(_amount);
+        require(depositTokenAmount > 0, "BaseStrategy::Withdraw amount too low");
+        uint256 withdrawAmount = _withdrawFromStakingContract(depositTokenAmount);
+        uint256 withdrawFee = _calculateWithdrawFee(depositTokenAmount);
+        depositToken.safeTransfer(msg.sender, withdrawAmount - withdrawFee);
+        _burn(msg.sender, _amount);
+        emit Withdraw(msg.sender, depositTokenAmount);
+    }
+
+    /**
+     * @notice Withdraw fee bips from underlying farm
+     * @dev Important: Do not override if withdraw fee is deducted from the amount returned by _withdrawFromStakingContract
+     */
+    function _getWithdrawFeeBips() internal view virtual returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @notice Calculate withdraw fee of underlying farm
+     * @dev Override if withdraw fee is calculated dynamically
+     * @dev Important: Do not override if withdraw fee is deducted from the amount returned by _withdrawFromStakingContract
+     */
+    function _calculateWithdrawFee(uint256 _amount) internal view virtual returns (uint256) {
+        uint256 withdrawFeeBips = _getWithdrawFeeBips();
+        return (_amount * withdrawFeeBips) / _bip();
+    }
+
+    function reinvest() external override {
+        _reinvest(false);
+    }
+
+    function _convertPoolRewardsToRewardToken() private returns (uint256) {
+        uint256 rewardTokenAmount = rewardToken.balanceOf(address(this));
+        uint256 count = supportedRewards.length;
+        for (uint256 i = 0; i < count; i++) {
+            address reward = supportedRewards[i];
+            if (reward == address(WAVAX)) {
+                uint256 balance = address(this).balance;
+                if (balance > 0) {
+                    WAVAX.deposit{value: balance}();
+                }
+                if (address(rewardToken) == address(WAVAX)) {
+                    rewardTokenAmount += balance;
+                    continue;
+                }
+            }
+            uint256 amount = IERC20(reward).balanceOf(address(this));
+            if (amount > 0) {
+                FormattedOffer memory offer = simpleRouter.query(amount, reward, address(rewardToken));
```

Comment:
Using yakSwapFallback is a valid SimpleRouter configuration. On a chain with high gas limit and cheap execution we might opt to use YakRouter exclusively and don't configure any "special" routes.

The dependency on SimpleRouter and central management of routes is on purpose. We do want to move away from hardcoded pairs/paths. Also note that the SimpleRouter instance is updatable in BaseStrategy.

If there is a path missing the strategy would malfunction but no rewards and more importantly user funds are at risk. We will test every strategy after deployment and missing configurations will be noticed quickly.

Reply:
Based on our experiments on L2s I wouldn't count on calling YakSwap queries from transaction. 

On Optimism the gas is incredibly cheap, but there is low gas limit, quickly reached by YakSwap findBestPath even with the more gas friendly configuration. 

On Arbitrum gas limit is higher, but unfortunately gas is much more expensive costing even $5 in tx fees for a simple query. 

Note that these are the results for only the gas-friendly configuration. Now, if users will want to use YakSwap for best price discovery, like on Avalanche now, we will have to enable more complex configurations. These may be costly even on L2s and even more likely won't be executable due to the gas limit. 
Of course, one could deploy their own YakSwap router with the simplest configuration to use as a fallback. But then it is up to you to maintain its list of adapters as liquidity sources change. 

To sum up, I believe it will be difficult to do generalised best path discovery within a transaction. I hope I am wrong, as this would enable many cool solutions.

Reply:
Existence of the path between reward and deposit token is essential for the strategy. Thus, it wouldn't hurt to be proactive and check whether it exists during initialisation.

Reply:
I now agree with centralised path storing. After all, paths between two tokens could be constantly updated and you would want to update it in inly one place.

Reply:
Was about to add the suggested check but we actually can't. Very often deposit token is not a simple ERC20 but an LP token where there will be no direct path from reward to deposit token. In those cases we override _convertRewardTokenToDepositToken and add liquidity on the underlying protocol.